packages/react-instantsearch-hooks-server/src/getServerState.tsx

```diff
@@ -153,3 +180,24 @@ function getInitialResults(rootIndex: IndexWidget): InitialResults {
 
   return initialResults;
 }
+
+function importRenderToString() {
+  return Promise.all([
+    // React pre-18 doesn't use `exports` in package.json, requiring a fully resolved path
+    // Thus, only one of these imports is correct
+    // eslint-disable-next-line import/extensions
+    import('react-dom/server.js').catch(() => {}),
+    import('react-dom/server').catch(() => {}),
```

Comment:
Can we test this by mocking the module with Jest?

Reply:
I didn't yet think of that, good call!

Reply:
I've been trying a lot of options, but at the end of the day, jest can't mock dynamic imports per test, meaning it would require a whole file per test case: https://github.com/facebook/jest/issues/10428. Here's some flavours of test I tried to write which don't pass if ran together:

```tsx
  describe('react-dom imports', () => {
    test('works when one of the two imports exists (no extension)', async () => {
      let mocked: { getServerState: typeof getServerState } = null!;
      jest.isolateModules(() => {
        jest.doMock('react-dom/server.js', () => {
          throw new Error('simulate the module not existing');
        });

        jest.doMock('react-dom/server', () => {
          return jest.requireActual('react-dom/server');
        });

        mocked = require('../getServerState');
      });

      const searchClient = createSearchClient({});
      const { App } = createTestEnvironment({ searchClient });

      const serverState = await mocked.getServerState(<App />);

      expect(serverState.initialResults).toEqual(expect.any(Object));
    });

    test('works when one of the two imports exists (with extension)', async () => {
      let mocked: { getServerState: typeof getServerState } = null!;
      jest.isolateModules(() => {
        jest.doMock('react-dom/server.js', () => {
          return jest.requireActual('react-dom/server');
        });

        jest.doMock('react-dom/server', () => {
          throw new Error('simulate the module not existing');
        });

        mocked = require('../getServerState');
      });

      const searchClient = createSearchClient({});
      const { App } = createTestEnvironment({ searchClient });

      const serverState = await mocked.getServerState(<App />);

      expect(serverState.initialResults).toEqual(expect.any(Object));
    });

    test('throws when no react-dom is present', async () => {
      let mocked: { getServerState: typeof getServerState } = null!;
      jest.isolateModules(() => {
        jest.doMock('react-dom/server.js', () => {
          throw new Error('simulate the module not existing');
        });

        jest.doMock('react-dom/server', () => {
          throw new Error('simulate the module not existing');
        });

        mocked = require('../getServerState');
      });

      const searchClient = createSearchClient({});
      const { App } = createTestEnvironment({ searchClient });

      await expect(
        mocked.getServerState(<App />)
      ).rejects.toThrowErrorMatchingInlineSnapshot(
        `"Could not import ReactDOMServer."`
      );
    });
  });
```

Reply:
I know, this is annoying. I'm OK with creating a file per test here.